diff --git a/microservices/dlstreamer-pipeline-server/plugins/camera/src-gst-gencamsrc/plugins/genicam.cc b/microservices/dlstreamer-pipeline-server/plugins/camera/src-gst-gencamsrc/plugins/genicam.cc
index 9e85c2d3..ea3c2829 100644
--- a/microservices/dlstreamer-pipeline-server/plugins/camera/src-gst-gencamsrc/plugins/genicam.cc
+++ b/microservices/dlstreamer-pipeline-server/plugins/camera/src-gst-gencamsrc/plugins/genicam.cc
@@ -4,6 +4,9 @@
 # SPDX-License-Identifier: Apache-2.0
 */
 
+#include <stdio.h>
+#include <time.h>
+#include <gst/gst.h>
 #include "genicam.h"
 #include "gstgencamsrc.h"
 
@@ -677,6 +680,28 @@ Genicam::Stop (void)
 }
 
 
+void
+Genicam::addReferenceTimestampMeta (GstBuffer * buf, guint64 timestampNS)
+{
+  // Convert nanoseconds to microseconds for gst_date_time_new_from_unix_epoch_utc_usecs.
+  GstDateTime *dt =
+      gst_date_time_new_from_unix_epoch_utc_usecs (timestampNS / 1000);
+  if (dt) {
+    gchar *dt_str = gst_date_time_to_iso8601_string (dt);
+    GST_DEBUG_OBJECT (gencamsrc,
+        "Current PTP Time: %s, %" G_GUINT64_FORMAT ", %" G_GUINT64_FORMAT,
+        dt_str, timestampNS / 1000);
+    g_free (dt_str);
+    gst_date_time_unref (dt);
+  }
+
+  GstCaps *ref_caps = gst_caps_new_empty_simple ("timestamp/x-hardware");
+  gst_buffer_add_reference_timestamp_meta (buf, ref_caps, timestampNS,
+      GST_CLOCK_TIME_NONE);
+  gst_caps_unref (ref_caps);
+}
+
+
 bool Genicam::Create (GstBuffer ** buf, GstMapInfo * mapInfo)
 {
   /* Grab the buffer, copy and release, set framenum */
@@ -718,6 +743,9 @@ bool Genicam::Create (GstBuffer ** buf, GstMapInfo * mapInfo)
       return FALSE;
     }
     GST_BUFFER_PTS (*buf) = timestampNS;
+
+    addReferenceTimestampMeta (*buf, timestampNS);
+
     gst_buffer_map (*buf, mapInfo, GST_MAP_WRITE);
 
     memcpy (mapInfo->data, buffer->getGlobalBase (), mapInfo->size);
diff --git a/microservices/dlstreamer-pipeline-server/plugins/camera/src-gst-gencamsrc/plugins/genicam.h b/microservices/dlstreamer-pipeline-server/plugins/camera/src-gst-gencamsrc/plugins/genicam.h
index c3f47f92..84d85be0 100755
--- a/microservices/dlstreamer-pipeline-server/plugins/camera/src-gst-gencamsrc/plugins/genicam.h
+++ b/microservices/dlstreamer-pipeline-server/plugins/camera/src-gst-gencamsrc/plugins/genicam.h
@@ -290,6 +290,9 @@ private:
 
   /* Sets Device Link Throughput Limit */
   bool setDeviceLinkThroughputLimit (void);
+
+  /* Logs frame timestamp and attaches reference timestamp metadata */
+  void addReferenceTimestampMeta (GstBuffer * buf, guint64 timestampNS);
 };
 
 #endif
