diff --git a/dlstreamer-pipeline-server/user_scripts/gvapython/sscape/sscape_adapter.py b/dlstreamer-pipeline-server/user_scripts/gvapython/sscape/sscape_adapter.py
index be9413a4..491ed19a 100644
--- a/dlstreamer-pipeline-server/user_scripts/gvapython/sscape/sscape_adapter.py
+++ b/dlstreamer-pipeline-server/user_scripts/gvapython/sscape/sscape_adapter.py
@@ -28,6 +28,7 @@ from sscape_policies import (
   ocrPolicy,
 )
 from sscape_3d_detector import Object3DChainedDataProcessor
+from gi.repository import Gst
 
 ROOT_CA = os.environ.get("ROOT_CA", "/run/secrets/certs/scenescape-ca.pem")
 DATETIME_FORMAT = "%Y-%m-%dT%H:%M:%S.%f"
@@ -278,10 +279,13 @@ class PostInferenceDataPublish:
       return tracker_id
     return len(existing_objects) + 1
 
-  def buildObjData(self, gvadata):
+  def buildObjData(self, gvadata, frame_ptp_timestamp):
     now = time.time()
+    if frame_ptp_timestamp is None:
+      frame_ptp_timestamp = gvadata['postdecode_timestamp']
+
     self.frame_level_data.update({
-      'timestamp': gvadata['postdecode_timestamp'],
+      'timestamp': frame_ptp_timestamp,
       'debug_timestamp_end': f"{datetime.fromtimestamp(now, tz=timezone(TIMEZONE)).strftime(DATETIME_FORMAT)[:-3]}Z",
       'debug_processing_time': now - float(gvadata['timestamp_for_next_block']),
       'rate': float(gvadata['fps'])
@@ -364,7 +368,25 @@ class PostInferenceDataPublish:
 
       if 'original_image_base64' in gvametadata:
         original_image_base64 = gvametadata['original_image_base64']
-      self.buildObjData(gvametadata)
+
+      ## For debugging timestamp meta from gst buffer
+      frame_ptp_timestamp = None
+      gst_buffer = frame._VideoFrame__buffer  # TODO: use public VideoFrame API when available
+      caps = Gst.Caps.from_string("timestamp/x-hardware")
+      if caps is not None:
+        try:
+          meta = gst_buffer.get_reference_timestamp_meta(caps)
+        except Exception as e:
+          print("Failed to read PTP reference timestamp meta: %s", e)
+          meta = None
+        if meta is not None:
+          frame_ptp_timestamp = f"{datetime.fromtimestamp(meta.timestamp/1e9, tz=timezone(TIMEZONE)).strftime(DATETIME_FORMAT)[:-3]}Z"
+          print("PTP META timestamp: %s", frame_ptp_timestamp)
+        else:
+          print("PTP META is not available in the gst buffer")
+      ## end of debugging timestamp meta
+
+      self.buildObjData(gvametadata, frame_ptp_timestamp)
 
       if self.is_publish_image:
         self.buildImgData(annotated_img, frame, True, original_image_base64)
