diff --git a/dlstreamer-pipeline-server/user_scripts/gvapython/sscape/sscape_adapter.py b/dlstreamer-pipeline-server/user_scripts/gvapython/sscape/sscape_adapter.py
index 47161724..e11d8c8e 100644
--- a/dlstreamer-pipeline-server/user_scripts/gvapython/sscape/sscape_adapter.py
+++ b/dlstreamer-pipeline-server/user_scripts/gvapython/sscape/sscape_adapter.py
@@ -271,7 +271,8 @@ class PostInferenceDataPublish:
       'timestamp': gvadata['postdecode_timestamp'],
       'debug_timestamp_end': f"{datetime.fromtimestamp(now, tz=timezone(TIMEZONE)).strftime(DATETIME_FORMAT)[:-3]}Z",
       'debug_processing_time': now - float(gvadata['timestamp_for_next_block']),
-      'rate': float(gvadata['fps'])
+      'rate': float(gvadata['fps']),
+      'frame': gvadata.get('sei_frame_num')
     })
     if 'initial_intrinsics' in gvadata:
       self.frame_level_data['initial_intrinsics'] = gvadata['initial_intrinsics']
