From 3864593b1fd19543e74e54e9b4c269e53e664a88 Mon Sep 17 00:00:00 2001
From: Lewis Liu <wei.zhi.liu@intel.com>
Date: Wed, 26 Feb 2025 17:43:38 +0800
Subject: [PATCH 03/14] add c++17 support (#4)

Signed-off-by: Liu, Wei Zhi <wei.zhi.liu@intel.com>
---
 CMakeLists.txt        | 7 ++++++-
 include/LoopClosing.h | 2 +-
 2 files changed, 7 insertions(+), 2 deletions(-)

diff --git a/CMakeLists.txt b/CMakeLists.txt
index 016e743..0c20133 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -14,9 +14,14 @@ set(CMAKE_CXX_FLAGS_RELEASE "${CMAKE_CXX_FLAGS_RELEASE} -march=native")
 
 # Check C++11 or C++0x support
 include(CheckCXXCompilerFlag)
+CHECK_CXX_COMPILER_FLAG("-std=c++17" COMPILER_SUPPORTS_CXX17)
 CHECK_CXX_COMPILER_FLAG("-std=c++11" COMPILER_SUPPORTS_CXX11)
 CHECK_CXX_COMPILER_FLAG("-std=c++0x" COMPILER_SUPPORTS_CXX0X)
-if(COMPILER_SUPPORTS_CXX11)
+if(COMPILER_SUPPORTS_CXX17)
+   set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=c++17")
+   add_definitions(-DCOMPILEDWITHC11)
+   message(STATUS "Using flag -std=c++17.")
+elseif(COMPILER_SUPPORTS_CXX11)
    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=c++11")
    add_definitions(-DCOMPILEDWITHC11)
    message(STATUS "Using flag -std=c++11.")
diff --git a/include/LoopClosing.h b/include/LoopClosing.h
index c9ccb26..7b6bfe0 100644
--- a/include/LoopClosing.h
+++ b/include/LoopClosing.h
@@ -223,7 +223,7 @@ protected:
     bool mbFixScale;
 
 
-    bool mnFullBAIdx;
+    int mnFullBAIdx;
 
 
 
-- 
2.30.2

