From 38e4c65bc734842f8f0cdcde4a22ff98730977cc Mon Sep 17 00:00:00 2001
From: Lewis Liu <wei.zhi.liu@intel.com>
Date: Thu, 20 Mar 2025 08:37:59 +0800
Subject: [PATCH 11/14] update inertialBA precondition checking. (#12)

Signed-off-by: Liu, Wei Zhi <wei.zhi.liu@intel.com>
---
 src/Optimizer.cc | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/src/Optimizer.cc b/src/Optimizer.cc
index b785be3..6b81220 100644
--- a/src/Optimizer.cc
+++ b/src/Optimizer.cc
@@ -2834,7 +2834,8 @@ void Optimizer::LocalInertialBA(KeyFrame *pKF, bool *pbStopFlag, Map *pMap, int&
     //cout << "Total map points: " << lLocalMapPoints.size() << endl;
     for(map<int,int>::iterator mit=mVisEdges.begin(), mend=mVisEdges.end(); mit!=mend; mit++)
     {
-        assert(mit->second>=3);
+        //fix-me: don't assert the size of observations. In stead, just skip if not matching.
+        //assert(mit->second>=3);
     }
 
     optimizer.initializeOptimization();
-- 
2.30.2

